src/experimental/patterns/ec2-app.ts

```diff
@@ -0,0 +1,282 @@
+ *
+ * It'll unset the `DesiredCapacity` of the ASG as a rolling update sets `Desired` back to the template version,
+ * undoing any changes that a scale-out event may have done.
+ * Having `DesiredCapacity` unset ensures the service remains at-capacity at all times.
+ *
+ * It'll also make the `MinInstancesInService` property dynamic via a CFN Parameter that Riff-Raff will set.
+ * The value depends on the current capacity of the ASG:
+ *  - If the service is running normally, it'll be set to the `MinSize` capacity
+ *  - If the service is partially scaled, it'll be set to the current `DesiredCapacity`
+ *  - If the service is fully scaled, it'll be set to (at least) `MaxSize` - 1
+ *
+ * @privateRemarks
+ * - Temporarily implemented as a singleton to ensure only one instance is added to a stack,
+ *   else multiple attempts to add the same CFN Parameter will be made and fail.
+ * - Once out of experimental, instantiate this `Aspect` directly in {@link GuStack}.
+ *
+ * @see https://github.com/guardian/testing-asg-rolling-update
+ */
+class HorizontallyScalingDeploymentProperties implements IAspect {
+  public readonly stack: Stack;
+  private static instance: HorizontallyScalingDeploymentProperties | undefined;
+
+  private constructor(scope: GuStack) {
+    this.stack = scope;
+  }
+
+  public static getInstance(stack: GuStack): HorizontallyScalingDeploymentProperties {
+    if (!this.instance || !isSingletonPresentInStack(stack, this.instance)) {
+      this.instance = new HorizontallyScalingDeploymentProperties(stack);
+    }
+
+    return this.instance;
+  }
+
+  public visit(construct: IConstruct) {
+    if (construct instanceof CfnScalingPolicy) {
+      const { node } = construct;
+      const { scopes, path } = node;
+      const guStack = GuStack.of(construct);
+
+      const autoScalingGroup = scopes.find((_): _ is GuAutoScalingGroup => _ instanceof GuAutoScalingGroup);
+
+      if (!autoScalingGroup) {
+        throw new Error(`Failed to detect the autoscaling group relating to the scaling policy on path ${path}`);
+      }
+
+      const cfnAutoScalingGroup = autoScalingGroup.node.defaultChild as CfnAutoScalingGroup;
+      const currentRollingUpdate = cfnAutoScalingGroup.cfnOptions.updatePolicy?.autoScalingRollingUpdate;
+
+      if (currentRollingUpdate) {
+        /*
+        An autoscaling group that horizontally scales should not explicitly set `Desired`,
+        as a rolling update will set the current `Desired` back to the template version,
+        undoing any changes that a scale-out event may have done.
+         */
+        cfnAutoScalingGroup.desiredCapacity = undefined;
+
+        const minInstancesInService = new CfnParameter(guStack, `MinInstancesInServiceFor${autoScalingGroup.app}`, {
+          type: "Number",
+          default: parseInt(cfnAutoScalingGroup.minSize),
+          maxValue: parseInt(cfnAutoScalingGroup.maxSize) - 1,
+        });
```

Comment:
If the `app` contains hyphens (`-`), or other illegal characters for a CFN Parameter, they'll get stripped out. This means Riff-Raff cannot reliably just find CFN Parameters that start with `MinInstancesInServiceFor`. Therefore Riff-Raff's logic would be something like this:
1. Find all `AWS::AutoScaling::ScalingPolicy`s in the CloudFormation template that has `MinInstancesInService` set via a CloudFormation Parameter
2. Find current `DesiredCapacity` of ASG that relates to the Scaling Policy
3. Set the CFN Parameter identified in step 1 to the value identified in step 2